cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+ * @param quotechar Quote character
+ * @param escapechar Delimiter escape character
+ * @param commentchar Comment line character (skip rows starting with this character)
+ * @param skipblanklines Skip empty rows
+ **/
+__global__ void __launch_bounds__(rowofs_block_dim) gather_row_offsets_gpu(uint64_t *row_ctx,
+                                                                           uint64_t *offsets_out,
+                                                                           const char *start,
+                                                                           size_t chunk_size,
+                                                                           size_t parse_pos,
+                                                                           size_t start_offset,
+                                                                           size_t data_size,
+                                                                           size_t byte_range_start,
+                                                                           size_t byte_range_end,
+                                                                           size_t skip_rows,
+                                                                           size_t num_row_offsets,
+                                                                           int terminator,
+                                                                           int delimiter,
+                                                                           int delimiter_esc,
+                                                                           int quotechar,
+                                                                           int escapechar,
+                                                                           int commentchar,
+                                                                           bool skipblanklines)
+{
+  __shared__ __align__(8) uint64_t ctxtree[rowofs_block_dim * 2];
+
+  const char *end = start + (min(parse_pos + chunk_size, data_size) - start_offset);
+  uint32_t t      = threadIdx.x;
+  size_t block_pos =
+    (parse_pos - start_offset) + blockIdx.x * static_cast<size_t>(rowofs_block_bytes) + t * 32;
+  const char *cur = start + block_pos;
+  uint64_t ctxb;
+  uint4 ctx_map;
+  int c, c_prev;
+
+  // Initial state is neutral context, zero rows
+  ctx_map.x = ctx_map.y = ctx_map.z = 0;
+  ctx_map.w                         = (0 << 0) | (1 << 2) | (2 << 4) | (3 << 6);  // i << (2*i)
+  c_prev                            = (cur > start) ? cur[-1] : terminator;
+  // Loop through all 32 bytes and keep a bitmask of row starts for each possible input context
+  for (uint32_t pos = 0; pos < 32; pos++, cur++, c_prev = c) {
+    uint32_t ctx;
+    if (cur < end) {
+      c = cur[0];
+      if (c_prev == terminator) {
+        if (c == commentchar ||
+            (skipblanklines && (c == terminator || (c == '\r' && terminator == '\n')))) {
```

Comment:
Question about `(c == '\r' && terminator == '\n')`. What if the character following the `\r` is not a `\n`? Wouldn't that cause `\r` to act the same as the comment escape char?

Reply:
Yes, I noticed that too in the original implementation, so assuming it's ok for now, but probably will need to be revisited when addressing pandas compatibility (it's certainly convenient for now since it doesn't require looking up the next character, which will almost certainly be needed for matching pandas).

Reply:
I suppose this would be an unusual edge case. Would a lone non-terminator `\r` or other control character at the start of a line be considered invalid CSV? On the other hand, if this is unusual but valid CSV, then it seems like a bug.

Reply:
I think the original implementation just assumes '\r' is always followed by '\n' (dos line endings), so it's almost certainly a bug unless pandas also does this.

Reply:
The blank line removal is now in its own count_if/remove_if block (can use some suggestions there if we can efficiently share the is_blank lambda).